### PR TITLE
Fixes complex datatype handling in ddp

### LINKED ISSUE #158753: DDP + Complex Dtype Appears to be Bugged
### 🐛 Describe the bug

Using a complex valued model converges on single GPU, but does not converge when using DDP. It appears that the imaginary component of the gradient is always zero with DDP. Wrapping in no_sync() and performing a manual all-reduce seems to circumvent this issue. There are three cases to observe in the following script 1) single GPU, 2) DDP, 3) DDP with manual reduce:

1) torchrun test.py
2) torchrun test.py --ddp
3) torchrun test.py --ddp --manual-reduce

Related issues:
#71613

```python
import torch
import torch.nn as nn
import torch.optim as optim
from torch.utils.data import Dataset, DataLoader, DistributedSampler
import torch.distributed as dist
import torch.multiprocessing as mp
import os
import argparse
import torch.nn.functional as F

class ComplexToyDataset(Dataset):
    def __init__(self, n_samples=10000, input_dim=4, noise_std=0.01):
        self.input_dim = input_dim

        real = torch.randn(n_samples, input_dim)
        imag = torch.randn(n_samples, input_dim)
        self.x = torch.complex(real, imag)  # Shape: (n_samples, input_dim)

        z1 = self.x[:, 0]
        z2 = self.x[:, 1]
        y = (z1 ** 2) + torch.sin(z2) + (0.5 * z1 * z2.conj())

        noise = torch.randn(n_samples, dtype=torch.cfloat) * noise_std
        self.y = (y + noise).unsqueeze(1)  # Shape: (n_samples, 1)

    def __len__(self):
        return len(self.x)

    def __getitem__(self, idx):
        return self.x[idx], self.y[idx]

class CModReLU(nn.Module):
    def __init__(self):
        super().__init__()
        self.b = nn.Parameter(torch.tensor(1e-5))

    def forward(self, x):
        return F.relu(torch.abs(x) + self.b) * torch.exp(1.j * torch.angle(x))

class ComplexLinear(nn.Module):
    def __init__(self, in_features, out_features):
        super().__init__()
        self.weight = nn.Parameter(torch.randn(out_features, in_features, dtype=torch.cfloat))
        self.bias = nn.Parameter(torch.randn(out_features, dtype=torch.cfloat))

    def forward(self, x):
        return torch.matmul(x, self.weight.t()) + self.bias

class ComplexModel(nn.Module):
    def __init__(self, input_dim=4):
        super().__init__()
        self.fc1 = ComplexLinear(input_dim, 8)
        self.fc2 = ComplexLinear(8, 1)
        self.act = CModReLU()

    def forward(self, x):
        x = self.act(self.fc1(x)) 
        return self.fc2(x)

def train(rank, world_size, use_ddp, manual_reduce):
    if use_ddp:
        dist.init_process_group("nccl", rank=rank, world_size=world_size)

    dataset = ComplexToyDataset()
    sampler = DistributedSampler(dataset, num_replicas=world_size, rank=rank) if use_ddp else None
    dataloader = DataLoader(dataset, batch_size=32, sampler=sampler, shuffle=False)

    model = ComplexModel()
    device = torch.device(f"cuda:{rank}" if torch.cuda.is_available() else "cpu")
    model.to(device)

    if use_ddp:
        model = nn.parallel.DistributedDataParallel(model, device_ids=None if not torch.cuda.is_available() else [rank])
    
    criterion = lambda y_pred, y_true: torch.mean(torch.abs(y_pred - y_true) ** 2)
    optimizer = optim.AdamW(model.parameters(), lr=1e-3)

    for epoch in range(100):
        grad_stats = {}
        running_loss = 0
        for batch_idx, (x, y) in enumerate(dataloader):
            x, y = x.to(device), y.to(device)
            optimizer.zero_grad()
            if use_ddp and manual_reduce:
                with model.no_sync():
                    out = model(x)
                    loss = criterion(out, y)
                    loss.backward()
            else:
                out = model(x)
                loss = criterion(out, y)
                loss.backward()

            if use_ddp and manual_reduce:
                for param in model.parameters():
                    if param.grad is not None:
                        dist.all_reduce(param.grad.data, op=dist.ReduceOp.SUM)

            for name, param in model.named_parameters():
                if torch.is_complex(param):
                    if param.grad is not None:
                        real_mean = param.grad.real.mean().item()
                        imag_mean = param.grad.imag.mean().item()
                        grad_stats.setdefault(name, []).append((real_mean, imag_mean))
            
            optimizer.step()
            running_loss += (loss.item() - running_loss) / (batch_idx + 1)

        if rank == 0 or not use_ddp:
            print(f"Epoch {epoch}, Running Loss: {running_loss:.4f}")
            print(f"\nEpoch {epoch} gradient stats:")
            for name, values in grad_stats.items():
                real_vals = [v[0] for v in values]
                imag_vals = [v[1] for v in values]
                avg_real = sum(real_vals) / len(real_vals)
                avg_imag = sum(imag_vals) / len(imag_vals)
                print(f"  {name}: avg grad real = {avg_real:.4e}, avg grad imag = {avg_imag:.4e}")

    if use_ddp:
        dist.destroy_process_group()

def main():
    parser = argparse.ArgumentParser()
    parser.add_argument("--ddp", action="store_true", help="Use DistributedDataParallel")
    parser.add_argument("--manual-reduce", action="store_true")
    parser.add_argument("--world_size", type=int, default=2, help="Number of processes for DDP")
    args = parser.parse_args()

    if args.ddp:
        mp.spawn(train, args=(args.world_size, True, args.manual_reduce), nprocs=args.world_size, join=True)
    else:
        train(rank=0, world_size=1, use_ddp=False, manual_reduce=False)

if __name__ == "__main__":
    main()
```

### Versions

(filtered, can provide more info)
[pip3] torch==2.7.1+cu118
[pip3] torchaudio==2.7.1+cu118
[pip3] torchvision==0.22.1+cu118
[pip3] triton==3.3.1

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta

### PR DESCRIPTION
**Pull request by arkadip-maitra**
Fixes #158753 

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @msaroufim @dcci

### CODE REVIEW

**Comment by arkadip-maitra** [comment]
@bdhirsh can you check this pr pls

**Comment by arkadip-maitra** [comment]
@wconstab @ezyang can you check this pr pls. thnx!!

File: torch/csrc/distributed/c10d/reducer.cpp [hunk 1]
```diff
@@ -1151,6 +1151,9 @@ void Reducer::initialize_buckets(
         }
         if (!options.has_dtype()) {
           options = options.dtype(variable.dtype());
+          if (variable.is_complex()) {
```
Comment (wconstab) [inline torch/csrc/distributed/c10d/reducer.cpp:1154]: what about the possibility of a mixed bucket?  i don't think your logic guards against it.  or am i missing it?

**Review by wconstab — COMMENTED** [review]
mostly lgtm, but i think it'd be good to add a test case for mixed complex + real params and its ok if we just error and say thats unsupported, or add validation to the bucket building code to avoid putting the two together

File: torch/csrc/distributed/c10d/reducer.cpp (hunk 1 above)
Comment (arkadip-maitra) [reply torch/csrc/distributed/c10d/reducer.cpp:1154]: added test. its handling real and complex params

File: test/distributed/test_c10d_nccl.py [hunk 2]
```diff
@@ -1997,6 +1997,228 @@ def _test_nccl_backend(
...
+                real_as_complex = torch.complex(
+                    real_branch, torch.zeros_like(real_branch)
+                )
+                return self.final_fc(complex_branch + real_as_complex)
+
+        torch.manual_seed(42 + self.rank)
+        model = MixedModule().to(device)
+        ref_model = MixedModule().to(device)
+
+        # 0.001 forces tiny buckets, stressing the group by dtype logic
```
Comment (wconstab) [inline test/distributed/test_c10d_nccl.py]: i disagree with this comment.  By forcing tiny bucket's, you'd actually be making things easy on the group-by-dtype logic.  The thing i'd rather see tested is forcing a large bucket.  This way, DDP would be more likely to try to group the complex+real parameters into the same bucket, and we'd actually stress the logic that keeps them separate.

File: torch/csrc/distributed/c10d/reducer.cpp (hunk 1 above)
Comment (wconstab) [reply torch/csrc/distributed/c10d/reducer.cpp:1154]: hmm, i'm still not convinced- can you explain how this code prevents mixing complex+real?  maybe it is working correctly but i'm just not seeing how

Comment (arkadip-maitra) [reply torch/csrc/distributed/c10d/reducer.cpp:1154]: The bucket assignment happens in [compute_bucket_assignment_by_size() ](https://github.com/arkadip-maitra/pytorch/blob/909f20b5ea7ccf6083cb40b19e372580a2415616/torch/csrc/distributed/c10d/reducer.cpp#L2212), which uses a [BucketKey](https://github.com/arkadip-maitra/pytorch/blob/909f20b5ea7ccf6083cb40b19e372580a2415616/torch/csrc/distributed/c10d/reducer.cpp#L2190) composed of (scalar_type, device) to group parameters:

```
auto key = BucketKey(tensor.scalar_type(), tensor.device());
auto& bucket = buckets[key];
```
The `torch.cfloat` has `ScalarType::ComplexFloat` and `torch.float32` has `ScalarType::Float`, they produce different keys and are placed into different bucket accumulators in the hash map. This separation happens at bucket assignment time, before any bucket is even created.

By the time `initialize_buckets()` runs, the `bucket_indices` it receives already guarantee that each bucket contains only one dtype. The `is_complex_bucket` flag then just records this fact for proper view handling later.

Modified the test with bucket size 100

File: test/distributed/test_c10d_nccl.py (hunk 2 above)
Comment (arkadip-maitra) [reply test/distributed/test_c10d_nccl.py]: yes, that is true. modified the test with bucket size 100

File: torch/csrc/distributed/c10d/reducer.cpp (hunk 1 above)
Comment (wconstab) [reply torch/csrc/distributed/c10d/reducer.cpp:1154]: Thanks for explaining, that makes sense.

**Review by wconstab — DISMISSED** [review]

**Review by wconstab — APPROVED** [review]
